lib/cfn-nag/custom_rules/EC2NetworkAclEntryIneffectiveDenyRule.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require 'cfn-nag/violation'
+require_relative 'base'
+
+class EC2NetworkAclEntryIneffectiveDenyRule < BaseRule
+  def rule_text
+    'NetworkACL Entry Deny rules should affect all CIDR ranges.'
+  end
+
+  def rule_type
+    Violation::WARNING
+  end
+
+  def rule_id
+    'W69'
+  end
+
+  def audit_impl(cfn_model)
+    violating_egress_entries = []
+    violating_ingress_entries = []
+    cfn_model.resources_by_type('AWS::EC2::NetworkAcl')
+             .select do |nacl|
+      violating_egress_entries = deny_does_not_cover_all_cidrs?(cfn_model, nacl.network_acl_egress_entries) ||
+                                 reused_ports?(cfn_model, nacl.network_acl_egress_entries)
+      violating_ingress_entries = deny_does_not_cover_all_cidrs?(cfn_model, nacl.network_acl_ingress_entries) ||
+                                  reused_ports?(cfn_model, nacl.network_acl_ingress_entries)
+    end
+
+    violating_egress_entries.map(&:logical_resource_id) + violating_ingress_entries.map(&:logical_resource_id)
+  end
+
+  private
+
+  def deny_does_not_cover_all_cidrs?(cfn_model, nacl_entries)
+    nacl_entry_resources = []
+    nacl_entries.select do |nacl_entry|
+      nacl_entry_resources << cfn_model.resource_by_id(nacl_entry)
```

Comment:
i would have expect the Nacl in cfn-model to have a collection of NaclEntry objects as opposed to a collection of logical resource ids (so the rule writer doesn't have to dereference them)

Reply:
Ok, I've been going through this for a little bit today.  It was based on the API gateway work but I agree it's a bit tedious this way.  Would this necessitate creating a separate NetworkAclEntry parser?  

In testing now with cfn-model, here's an example NetworkAclEntry resource item as it stands:

`[#<AWS::EC2::NetworkAclEntry:0x0000556bb0856e40 @cfn_model=#<CfnModel:0x0000556bb0861980 @parameters={}, @resources={"testvpc1"=>#<AWS::EC2::VPC:0x0000556bb08611b0 @cfn_model=#<CfnModel:0x0000556bb0861980 ...>, @logical_resource_id="testvpc1", @resource_type="AWS::EC2::VPC", @metadata=nil, @cidrBlock="10.0.0.0/16", @enableDnsSupport=true, @enableDnsHostnames=true, @instanceTenancy="default">, "myNetworkAcl"=>#<AWS::EC2::NetworkAcl:0x0000556bb0857430 @cfn_model=#<CfnModel:0x0000556bb0861980 ...>, @network_acl_egress_entries=[], @network_acl_ingress_entries=[], @resource_type="AWS::EC2::NetworkAcl", @vpcId="testvpc1", @logical_resource_id="myNetworkAcl", @metadata=nil>, "EgressEntry1"=>#<AWS::EC2::NetworkAclEntry:0x0000556bb0856e40 ...>}, @conditions={}, @globals={}, @raw_model={"Resources"=>{"testvpc1"=>{"Type"=>"AWS::EC2::VPC", "Properties"=>{"CidrBlock"=>"10.0.0.0/16", "EnableDnsSupport"=>true, "EnableDnsHostnames"=>true, "InstanceTenancy"=>"default"}}, "myNetworkAcl"=>{"Type"=>"AWS::EC2::NetworkAcl", "Properties"=>{"VpcId"=>"testvpc1"}}, "EgressEntry1"=>{"Type"=>"AWS::EC2::NetworkAclEntry", "Properties"=>{"NetworkAclId"=>{"Ref"=>"myNetworkAcl"}, "Protocol"=>"6", "RuleAction"=>"allow", "RuleNumber"=>"100", "CidrBlock"=>"10.0.0.0/16", "Egress"=>true, "PortRange"=>{"From"=>"443", "To"=>"443"}}}}}, @line_numbers={}>, @logical_resource_id="EgressEntry1", @resource_type="AWS::EC2::NetworkAclEntry", @metadata=nil, @networkAclId={"Ref"=>"myNetworkAcl"}, @protocol="6", @ruleAction="allow", @ruleNumber="100", @cidrBlock="10.0.0.0/16", @egress=true, @portRange={"From"=>"443", "To"=>"443"}>]
`
This is including more than is desired.

Reply:
yea gw was a tough one.... the relationships among the model elements (on the aws side) don't make obvious sense and some are implicit.  a good rule though is to build any "referencing" or synthesizing into the underlying cfn-model so that rule developer can just manipulate objects/properties

Reply:
Do you feel it still helps to create separate ingress and egress objects for NACL since they seem to generally have separate rule applications?  Or does it make sense to offer a third option that includes all of them?

Reply:
i think you can have a single Entry object that covers both of them.  given it's an object you can hide some of the "statefulness" behind human readable predicates too e.g. is_egress?

Reply:
Updated